test/xds/xds_client_custom_dialopts_test.go

```diff
@@ -0,0 +1,212 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xds_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync/atomic"
+	"testing"
+
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/internal/stubserver"
+	"google.golang.org/grpc/internal/testutils"
+	"google.golang.org/grpc/internal/testutils/xds/e2e"
+	internalbootstrap "google.golang.org/grpc/internal/xds/bootstrap"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/stats"
+	"google.golang.org/grpc/xds/bootstrap"
+
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+)
+
+const testCredsBuilderName = "test_dialer_creds"
+
+// testCredsBuilder implements the `Credentials` interface defined in
+// package `xds/bootstrap` and encapsulates an insecure credential with a
+// custom Dialer that specifies how to dial the xDS server.
+type testCredsBuilder struct {
+	dialerCalled     atomic.Bool
+	tagRPCCalled     atomic.Bool
+	handleRPCCalled  atomic.Bool
+	tagConnCalled    atomic.Bool
+	handleConnCalled atomic.Bool
+}
+
+func (t *testCredsBuilder) Build(config json.RawMessage) (credentials.Bundle, func(), error) {
+	cfg := &struct {
+		MgmtServerAddress string `json:"mgmt_server_address"`
+	}{}
+	if err := json.Unmarshal(config, &cfg); err != nil {
+		return nil, func() {}, fmt.Errorf("failed to unmarshal config: %v", err)
+	}
+	return &testCredsBundle{insecure.NewBundle(), &t.dialerCalled, cfg.MgmtServerAddress, &noopStatsHandler{
+		tagRPCCalled:     &t.tagRPCCalled,
+		handleRPCCalled:  &t.handleRPCCalled,
+		tagConnCalled:    &t.tagConnCalled,
+		handleConnCalled: &t.handleConnCalled,
+	}}, func() {}, nil
+}
+
+func (t *testCredsBuilder) Name() string {
+	return testCredsBuilderName
+}
+
+// noopStatsHandler implements `stats.Handler`. It's a no-op mock handler.
+type noopStatsHandler struct {
+	tagRPCCalled     *atomic.Bool
```

Comment:
What would be the reason for using `*atomic.Bool` instead of just `atomic.Bool`?
Also see below comment regarding using `sync/atomic` package vs channel

Reply:
I believe the way things are setup instantiates atomics in `testCredsBuilder`, then I saw no reason to copy. Or anything I missed?